Source/Core/Core/HW/EXI/BBA/TAPServerConnection.cpp

```diff
@@ -0,0 +1,362 @@
+  m_read_enabled.Set();
+}
+
+void TAPServerConnection::RecvStop()
+{
+  m_read_enabled.Clear();
+}
+
+bool TAPServerConnection::SendAndRemoveAllHDLCFrames(std::string* send_buf)
+{
+  while (!send_buf->empty())
+  {
+    const std::size_t start_offset = send_buf->find(0x7E);
+    if (start_offset == std::string::npos)
+    {
+      break;
+    }
+    const std::size_t end_sentinel_offset = send_buf->find(0x7E, start_offset + 1);
+    if (end_sentinel_offset == std::string::npos)
+    {
+      break;
+    }
+    const std::size_t end_offset = end_sentinel_offset + 1;
+    const std::size_t size = end_offset - start_offset;
+
+    const u8 size_bytes[2] = {static_cast<u8>(size), static_cast<u8>(size >> 8)};
+    if (send(m_fd, reinterpret_cast<const char*>(size_bytes), 2, SEND_FLAGS) != 2)
+    {
+      ERROR_LOG_FMT(SP1, "SendAndRemoveAllHDLCFrames(): could not write size field");
+      return false;
+    }
+    const int written_bytes =
+        send(m_fd, send_buf->data() + start_offset, static_cast<int>(size), SEND_FLAGS);
+    if (u32(written_bytes) != size)
+    {
+      ERROR_LOG_FMT(SP1,
+                    "SendAndRemoveAllHDLCFrames(): expected to write {} bytes, instead wrote {}",
+                    size, written_bytes);
+      return false;
+    }
+    *send_buf = send_buf->substr(end_offset);
+  }
+  return true;
+}
+
+bool TAPServerConnection::SendFrame(const u8* frame, u32 size)
+{
+  INFO_LOG_FMT(SP1, "SendFrame {}\n{}", size, ArrayToString(frame, size, 0x10));
+
+  // On Windows, the data pointer is of type const char*; on other systems it is
+  // of type const void*. This is the reason for the reinterpret_cast here and
+  // in the other send/recv calls in this file.
+  const u8 size_bytes[2] = {static_cast<u8>(size), static_cast<u8>(size >> 8)};
+  if (send(m_fd, reinterpret_cast<const char*>(size_bytes), 2, SEND_FLAGS) != 2)
+  {
+    ERROR_LOG_FMT(SP1, "SendFrame(): could not write size field");
+    return false;
+  }
+  const int written_bytes =
+      send(m_fd, reinterpret_cast<const char*>(frame), static_cast<ws_ssize_t>(size), SEND_FLAGS);
+  if (u32(written_bytes) != size)
+  {
+    ERROR_LOG_FMT(SP1, "SendFrame(): expected to write {} bytes, instead wrote {}", size,
+                  written_bytes);
+    return false;
+  }
+  return true;
+}
+
+void TAPServerConnection::ReadThreadHandler()
+{
+  enum class ReadState
+  {
+    SIZE,
+    SIZE_HIGH,
+    DATA,
+    SKIP,
+  };
+  ReadState read_state = ReadState::SIZE;
+
+  std::size_t frame_bytes_received = 0;
+  std::size_t frame_bytes_expected = 0;
+  std::string frame_data;
+
+  while (!m_read_shutdown.IsSet())
+  {
+    fd_set rfds;
+    FD_ZERO(&rfds);
+    FD_SET(m_fd, &rfds);
+
+    timeval timeout;
+    timeout.tv_sec = 0;
+    timeout.tv_usec = 50000;
+    int select_res = select(m_fd + 1, &rfds, nullptr, nullptr, &timeout);
+    if (select_res < 0)
+    {
+      ERROR_LOG_FMT(SP1, "Can\'t poll tapserver fd: {}", Common::StrNetworkError());
+      break;
```

Comment:
You shouldn't break of the `while` loop, otherwise the `ReadThreadHandler` will finish and can't be used anymore.

If there is an error, this might suggest that the connection was closed/interrupted with the TAP server. You might want to create a `Reconnect` method which will close the failing socket and try to reconnect (maybe x times?) to the destination. If the reconnect method fails, you might want to close the socket and set the fd to -1.

Moreover, you might want to add the following:
 - Check that `m_read_enabled` is set before handling the socket:
```c++
  while (!m_read_shutdown.IsSet())
  {
    if (!m_read_enabled.IsSet())
      continue;
```
 - Then, reconnect to the TAP server if needed in `RecvStart`:
```c++
if (m_read_enabled.IsSet() && !IsActivated())
  ReconnectToDestination(m_destination);
```

Reply:
There is a good reason why we don't attempt to reconnect. The tapserver connection includes state on the remote end of the connection, so reconnecting isn't as simple as just opening the TCP connection again; we would also need to tell the remote end what its virtual network state should be, and it would have to reopen the downstream connections that were part of the original virtual network. Those connections may also have had server-side state of their own which was lost (in the case of PSO, they definitely would have), so reconnecting after a disconnection simply isn't feasible even if we were to reconstruct the remote tapserver's state perfectly.

Reply:
This still needs no be addressed as by doing nothing the game will still believe it is connected while it isn't. My comment regarding `m_read_enabled` and not breaking from the while loop still apply _(which might be responsible for the reconnection issue mentioned by @Matt-Swift)_.

Some questions regarding the error handling:
- What does the TAP server do when there is a client disconnection? _(i.e. does it notify the connected peers like the PSO server?)_
- What does the TAP server do when the client logs out and disconnects gracefully? _(i.e. does it close the TAP server connection and does the TAP server disconnect the client from its peers?)_
- What does the (PSO) game server do when a client disconnects? _(e.g. from a direct connection or a TAP server)_
- What does the (PSO) game server do when a client is trying to connect twice? _(i.e. would both be successful, would the first connection be disconnected or would the last connection raise an error?)_

Is the TAP server used open-source? Unless I'm mistaken, if it acts just like a regular TAP, trying to reconnect to it when there was a socket error (which probably means we got disconnected from the TAP server) shouldn't be that harmful. It either retained its previous state (which might be fine) or it doesn't and will try to create a new one (which should be fine as long as we notified the game that we were disconnected).

Reply:
[tapserver](https://github.com/fuzziqersoftware/tapserver) is a program that creates a "fake ethernet" interface when an inbound connection is received, and uses that to forward and receive traffic through the kernel's network stack. it does not act like a regular tap interface; it's designed to be similar, but there are some semantic differences. Notably, when the tapserver connection is closed, the associated network interface is deleted, which causes the kernel to close any TCP connections that were connected through it.

[newserv](https://github.com/fuzziqersoftware/newserv) implements the same protocol as tapserver, but implements its own Ethernet, PPP, IP, and TCP stacks internally, which does a better job of handling some of the quirks of PSO than just using tapserver and the modern kernel's network stack directly. In both projects, all TCP connections tunneled through the tapserver connection are closed when the tapserver connection is closed.

To answer your specific questions,
* When the client disconnects either normally or unexpectedly, tapserver does disconnect upstream TCP connections.
* When a client disconnects from the game server, the server removes the client from the lobby it was in (if any), notifies other clients in that lobby, deletes any game state associated with the player, and saves their character files if necessary.
* When a client connects twice simultaneously, the server accepts both connections, and these connections do not share any state (they may use different characters and be in different lobbies/games). newserv, at least, is intentionally designed to allow multiple logins on the same account at the same time, thought most other PSO servers don't allow this.

What do the other network interfaces do when their underlying links go down? If an actual tap interface (not tapserver) becomes unavailable, or an XLink socket can't receive for some reason, it looks like those BBA types would just loop forever trying to receive and logging error messages. Furthermore, it looks like the BBA NetworkInterface abstraction doesn't have any way to notify CEXIETHERNET that the link is down due to an external failure. This is not to say we shouldn't handle that case in the modem adapter, but the point of this PR is to bring its implementation up to speed with how the BBA currently works, and I believe that has been done. Figuring out how to notify the game that the interface is down due to an external failure will require significantly more reverse-engineering work, and I am not willing to spend an arbitrary amount of time making this perfect when it's already at parity with the other existing implementations.

Reply:
I admit that while this issue isn't perfect, I would be quite interested in seeing this implemented sooner rather than later, as ironically, the current BBA HLE that is implemented in Dolphin is not actually usable with Phantasy Star Online Episode I & II as the game will crash while online after 30 minutes to hour without fail (this is a separate issue that needs reporting if it hasn't already); this tapserver implementation on the other hand works flawlessly, and having to quickly restart the game in a non-critical spot is a major improvement over the current experience.

The reported issue of randomly disconnecting may have just been a fluke from my end, as a friend and I have been playing using this version of Dolphin without any issues recently on my own public newserv instance, but most importantly without any client freezes.

The most crucial addition of course though as well is the modem adapter, making Episode I & II Trial Edition accessible to many more players, and from my cursory testing with the modem adapter, there doesn't seem to be any unique issues with it.

Reply:
I do get your point but the current implementation isn't on par with the Win32 TAP implementation nor Unix implementation.

@fuzziqersoftware 
That's why I suggested to, at least, address the following issues:
- `m_read_enabled` needs to be checked before handling reads.
- Not exiting the while loop if there is an error on `select`, you can use `continue` instead _(like what the TAP Unix code uses to do)_ and add an error log.

@Matt-Swift 
Regarding the BBA HLE issue, there is an issue for it:
 - https://bugs.dolphin-emu.org/issues/13032

However, the lack of reproducibility, logs and the fact I don't own the game, that makes it very difficult to diagnose the root cause. My best attempt to fix the issue is this PR:
 - https://github.com/dolphin-emu/dolphin/pull/12560

As I don't own the game, I'm unsure if it solves the issue. Regardless, I also believe that tunneling the traffic to a TAP server is probably faster than emulating sending/receiving frames.

Reply:
I'm still not convinced that repeating select() after it has failed once is the correct thing to do (here or in the BBA implementation), but this PR has taken long enough already and I'm tired, so I've changed it as you suggested.